src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,1531 @@
+        elif algorithm == "kaltofen_shoup":
+            fmpz_mod_poly_factor_kaltofen_shoup(fac, self.val, self.ctx.mod.val)
+        elif algorithm == "berlekamp":
+            fmpz_mod_poly_factor_berlekamp(fac, self.val, self.ctx.mod.val)
+        else:
+            raise ValueError("unknown algorithm")
+        
+        res = [0] * fac.num
+
+        cdef fmpz_mod_poly u
+        for i in range(fac.num):
+            u = fmpz_mod_poly.__new__(fmpz_mod_poly)
+            u.ctx = self.ctx
+            fmpz_mod_poly_set(u.val, &fac.poly[i], self.ctx.mod.val)
+            exp = fac.exp[i]
+            res[i] = (u, exp)
+        return self.leading_coefficient(), res
+
+    def roots(self, multiplicities=True):
+        r"""
+        Return the roots of the polynomial in :math:`(\mathbb{Z}/N\mathbb{Z})^*`
+        Requires that the modulus `N` is prime.
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> x = R.gen()
+            >>> f = (x - 1) * (x - 2)**3 * (x - 3)**5
+            >>> f.roots()
+            [(fmpz_mod(1, 163), 1), (fmpz_mod(2, 163), 3), (fmpz_mod(3, 163), 5)]
+            >>> f.roots(multiplicities=False)
+            [fmpz_mod(1, 163), fmpz_mod(2, 163), fmpz_mod(3, 163)]
+        """
+        cdef fmpz_mod_poly_factor_t fac
+        cdef int i, with_multiplicity
+
+        if multiplicities:
+            with_multiplicity = 1
+
+        if not self.ctx.is_prime():
+            raise NotImplementedError("factor algorithm assumes that the modulus is prime")
+                 
+        fmpz_mod_poly_factor_init(fac, self.ctx.mod.val)
+        fmpz_mod_poly_roots(fac, self.val, with_multiplicity, self.ctx.mod.val)
+        res = [0] * fac.num
+
+        cdef fmpz_mod root
+        for i in range(fac.num):
+            root = fmpz_mod.__new__(fmpz_mod)
+            root.ctx = self.ctx.mod
+            fmpz_mod_poly_get_coeff_fmpz(
+                root.val, &fac.poly[i], 0, self.ctx.mod.val
+            )
+            fmpz_mod_neg(
+                root.val, root.val, root.ctx.val
+            )
+            if multiplicities:
+                mul = fac.exp[i]
+                res[i] = (root, mul)
+            else:
+                res[i] = root
+        return res
+
+    def complex_roots(self):
+        """
+        This method is not implemented for polynomials in
+        :math:`(\mathbb{Z}/N\mathbb{Z})[X]`
+        """
+        raise DomainError("Cannot compute compex roots for polynomials over integers modulo N")
+
+cdef class BerlekampMassey:
+    def __cinit__(self):
+        self.initialized = False
+
+    def __dealloc__(self):
+        if self.initialized:
+            fmpz_mod_berlekamp_massey_clear(self.B, self.ctx.mod.val)
+
+    def __init__(self, ctx):
+        if not typecheck(ctx, fmpz_mod_poly_ctx):
+            raise TypeError
+        self.ctx = ctx
+        fmpz_mod_berlekamp_massey_init(self.B, self.ctx.mod.val)
+        self.initialized = True
```

Comment:
Using `initialized` like this means that we don't need `__init__` any more. The code in `__init__` can be added to `__cinit__`.